### DISCUSSION TITLE
Long running processes

### DISCUSSION THREAD

**Discussion by StevePorter92** [NativePHP/laravel#178] [Ideas]
Hi folks 👋 !

Has any thought been put in to how we'd go about running long running processes or other background commands? 

If I'm understanding whats happening correctly, it looks like the electron app is starting the task scheduler and queue worker when bootstrapping NativePHP https://github.com/NativePHP/electron-plugin/blob/main/src/server/index.ts?

Could we look to do something similar to run other daemon-like commands or processes?

Would love to help contribute if possible.

**Comment by StevePorter92** [comment]
FYI, the use case is to run a task that keeps a raw socket connection open and responds to messages and broadcast events to update the UI

**Reply by eugenefvdm** [reply]
Hi @StevePorter92 ,

Please a look at this page, the very last paragraphs:
https://nativephp.com/docs/1/digging-deeper/broadcasting#broadcasting-custom-events

> This is great for times when you want to offload an intensive task to a background queue and await its completion without constantly polling your application for its status.
>
> Your fired event will be broadcast and your application can listen for it and just as your normally would.


It appears what you want is briefly covered here and what's further interesting more up on the page:


> NativePHP automatically runs php artisan websockets:serve to start your WebSocket server once your application runs, allowing you to use Laravel Echo to connect to it and listen for events in real-time.

So I have a similar request. My API needs to do a few background API calls to fetch data in SQLite for [Sushi](https://github.com/calebporzio/sushi). These are longer running processes so I can't burden the UI with it as I believe it could freeze. 

In theory this is all asynchronous Windows UI programming whereby threads are normally relevant but if I can just figure out how to queue Laravel tasks via a job or something I would be more than happy. Something like:

```php
FetchUsersFromApi::dispach();
```

If I can just figure out how to offload simple API tasks then I would be good to go. It seems the facility is there but my skill set just can't place it. I've checked and some say you can use SQLite with a Laravel queue worker, but how to start said queue worker? Or am I off track and should be doing this via Websockets?

My next steps are to get some kind of basic Websocket implementation going as per the documentation and then I'll update this post. Thanks for your post, I really appreciate the question as I'm in the same vein.

Please let me know how you progress :-)

**Reply by eugenefvdm** [reply]
I gave this part of the manual a go to get async tasks going:
https://nativephp.com/docs/1/digging-deeper/broadcasting

It is mentioned two ways to do it, using Websockets with it's overhead and complexities, and a simpler way using Electron IPC mechanisms.

In both instances I struggled. In the case of Websockets, it's just too complex for me with Echo, Pusher compatibility and Websockets. Too many technologies and quickly hard to accept defaults. I believe someone who knows Laravel Websockets and Pusher should be able to get it going.

In the case of the Electron IPC stuff, I quickly ran into this:

>  Vite does not automatically polyfill Node.js modules.

That error I got on the front-end when I tried to add the code below to app.js:

```javascript
import * as remote from '@electron/remote'
 
ipcRenderer.on('native-event', (_, data) => {
    const eventClassName = data.event;
    const eventPayload = data.payload;
});
```

Additional information:

> Module "path" has been externalized for browser compatibility. Cannot access "path.join" in client code

**Reply by eugenefvdm** [reply]
After much fiddling I got Websockets, Pusher, and Echo to work. I basically followed this [Gist](https://gist.github.com/codewithgun/140cfe807cfaf3cfbbeffbaeebf0add5) to fool Pusher that's it's local.

After having done all that, I realized I'd only gained the ability to launch async Javascript, which is not exactly what I wanted.

Then I discovered that actually Native PHP already has queueing functionality built-in but it's just not [documented](https://github.com/NativePHP/laravel/issues/97).

So in the end, this just works without any configuration:

```php
FetchUsersFromApi::dispatch();
```

Aloha Async Tasks! :-)

**Comment by osc2nuke** [comment]
Is your finding related to what i experienced (keep in mind, my laptop has very low specs)?
https://github.com/NativePHP/laravel/discussions/172



![WhatsApp Image 2023-08-14 at 20 44 16](https://github.com/NativePHP/laravel/assets/5169035/d1f8c53e-ca45-4673-8aca-522952d7e23c)

**Reply by StevePorter92** [reply]
I think we're talking about 2 different things. I'm interested in how we can go about running asynchronous php processes. 

It sounds like you've got an orphaned task problem.

**Comment by simonhamp** [comment]
I've moved this to Ideas because I think there's a valid case for this as a feature, but NativePHP doesn't currently support this.

First pass at this could be something in the NativePHP service provider that instructs Electron to spawn processes using the built-in PHP. Then as long as your script/Command that handles that execution runs in a loop, you could do whatever you want.

e.g. in the service provider:

```php
public function startupProcesses()
{
    return [
        'artisan socket:listen -p 4444',
    ];
}
```

In this, I'm imagining a Laravel Command class with the signature `socket:listen {-p}` that contains something like a ReactPHP loop that keeps a socket open listening for connections.

What do you think?

**Comment by SRWieZ** [comment]
This is a great idea!

In one of my projects, I have a file watcher that could use this. 

Ideally it would work like supervisord and restart the process if any error occurs. 

It would also be nice to start/stop/restart/status from other part of the code. 

In my example, the file watcher is started only after the user has configured the working directory. And restart it whenever it changes it.

**Comment by simonhamp** [comment]
This got [raised again](https://github.com/orgs/NativePHP/discussions/371) and is now [in active development](https://github.com/orgs/NativePHP/discussions/371#discussioncomment-11009046)!

**Comment by simonhamp** [comment]
Our Child Processes functionality [got released](https://github.com/NativePHP/laravel/releases/tag/0.6.0) yesterday!
